cpp/unreal_plugins/SpEngine/Source/SpEngine/WorkQueue.h

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <concepts>    // std::same_as
+#include <cmath>       // std::nan
+#include <future>
+#include <map>
+#include <mutex>
+#include <string>
+#include <utility>     // std::forward, std::move
+#include <vector>
+
+#include <SpCore/Assert.h>
+#include <SpCore/Boost.h>
+
+template <typename TFunc, typename TReturn, typename... TArgs>
+concept CCallable = requires(TFunc func, TArgs... args) {
+        { func(args...) } -> std::same_as<TReturn>;
+};
+
+
+template <typename TClass>
```

Comment:
I thought you could get rid of `FuncInfo` entirely. Did that end up not working for some reason?